File: c/src/gendataset.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <fftw3.h>

#define ASSERT_ALLOC(cmd) if(NULL == (cmd)) { printf("Not enough memory\n"); exit(EXIT_FAILURE); }

/* external modules */
extern int load_monopage_tiff_into_fftw_complex(fftw_complex**,int*,int*,char*,char); // see source file 'tiffread.c'
extern int save_monopage_striped_tiff_from_fftw_complex(char*,fftw_complex*,int,int,char,int,char*,char*,char*); // see source file 'tiffwritecomplex.c'
extern int getasciitranslations(double**,double**,int*,char*,char); // see source file 'ascii.c'
extern int gendataset(fftw_complex*, fftw_complex***, fftw_complex**, double*, double*, int*, int*, int*, int*, int, double*, double*, int, int, int, int, char, char); // see source file 'gendataset_kernel.c'

/* usage displayer */
static void display_usage()
{
  printf("\nCompute a realistic low-resolution stack from a high resolution image (avoiding periodic-like boundaries).\n\n");
  printf("Usage: gendataset [-ftype type] [-zx zx] [-zy zy] [-v] [-r ref] in T out\n\n");
  printf(" -ftype type : (default float) datatype of the output TIFF images, possible\n");
  printf("               choices are {uint8,int8,uint16,int16,uint32,int32,float}\n");
  printf(" -cx cx      : (default 10) maximal number of columns to be discarded from\n");
  printf("               the input image during the first cropping step (int >= 0))\n");
  printf(" -cy cy      : (default 10) maximal number of rows to be discarded from\n");
  printf("               the input image during the first cropping step (int >= 0))\n");
  printf(" -p p        : (default 5) the final high-resolution ROI size will be equal\n");
  printf("               to (q/p) * the dimensions of the cropped HR image (int >= 1)\n");
  printf(" -q q        : (default 4) see above (int in [1,p-1])\n");
  printf(" -zx zx      : (default 2.) X-axis subsampling factor (double >= 1.)\n");
  printf(" -zy zy      : (default 2.) Y-axis subsampling factor (double >= 1.)\n");
  printf(" -v          : enable verbose mode\n");
  printf(" -r ref      : output high resolution reference image (cropping of in)\n");
  printf(" in          : input high-resolution TIFF image (large domain)\n");
  printf(" T           : input sequence of displacements (ASCII format)\n");
  printf(" out         : output realistic sequence of low-resolution images (TIFF-stack)\n");
  printf("               with restricted domain to get rid of periodization artifacts\n\n");
}

/* command line interface */
int main(int argc, char **argv)
{
  char *datatype=NULL,*fname_in=NULL,*fname_T=NULL,*fname_out=NULL,*fname_ref=NULL,*zx_value=NULL,*zy_value=NULL,*p_value=NULL,*q_value=NULL,*cx_value=NULL,*cy_value=NULL,vflag=0;
  double *dx=NULL,*dy=NULL,zx=2.,zy=2.;
  fftw_complex **u0=NULL,*utmp=NULL,*ref=NULL;
  int err,k,type,nx,ny,nimages,ntranslations,Nx,Ny,cx=10,cy=10,p=5,q=4;

  /*************************/
  /* parse input arguments */
  /*************************/
  for(k=1; k<argc; k++) {
    if (strcmp(argv[k],"-v") == 0) vflag = 1;
    else if (strcmp(argv[k],"-ftype") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -ftype requires an argument\n\n"); return EXIT_FAILURE; }
      else { datatype = argv[k+1]; k++; }
    }
    else if (strcmp(argv[k],"-zx") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -zx requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	zx_value = argv[k+1]; err = sscanf(zx_value,"%lf",&zx); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -zx\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-zy") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -zy requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	zy_value = argv[k+1]; err = sscanf(zy_value,"%lf",&zy); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -zy\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-p") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -p requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	p_value = argv[k+1]; err = sscanf(p_value,"%d",&p); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -p\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-q") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -q requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	q_value = argv[k+1]; err = sscanf(q_value,"%d",&q); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -q\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-cx") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -cx requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	cx_value = argv[k+1]; err = sscanf(cx_value,"%d",&cx); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -cx\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-cy") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -cy requires an argument\n\n"); return EXIT_FAILURE; }
      else {
	cy_value = argv[k+1]; err = sscanf(cy_value,"%d",&cy); k++;
	if(err != 1) { display_usage(); printf("Error: could not retrieve properly the argument of option -cy\n\n"); return EXIT_FAILURE; }
      }
    }
    else if (strcmp(argv[k],"-r") == 0) {
      if(k==argc-1) { display_usage(); printf("Error: option -r requires an argument\n\n"); return EXIT_FAILURE; }
      else { fname_ref = argv[k+1]; k++; }
    }
    else if (NULL == fname_in) fname_in = argv[k];
    else if (NULL == fname_T) fname_T = argv[k];
    else if (NULL == fname_out) fname_out = argv[k];
    else { display_usage(); printf("Error: could not retrieve properly input arguments.\n\n"); return EXIT_FAILURE; }
  }
  
  /*********************/
  /* check consistency */
  /*********************/
  if ((NULL == fname_in) || (NULL == fname_T) || (NULL == fname_out)) {
    display_usage();
    if (NULL == fname_in) printf("Error: input 'in' is missing\n\n");
    else if (NULL == fname_T) printf("Error: input 'T' is missing\n\n");
    else printf("Error: output 'out' is missing\n\n");
    return EXIT_FAILURE;
  }

  if((NULL == datatype)||(strcmp(datatype,"float")==0)) type = 0;
  else if (strcmp(datatype,"uint8")==0) type = 8;
  else if (strcmp(datatype,"int8")==0) type = -8;
  else if (strcmp(datatype,"uint16")==0) type = 16;
  else if (strcmp(datatype,"int16")==0) type = -16;
  else if (strcmp(datatype,"uint32")==0) type = 32;
  else if (strcmp(datatype,"int32")==0) type = -32;
  else { display_usage(); printf("Error: unrecognized value for option -ftype.\n\n"); return EXIT_FAILURE; }

  if(zx < 1.) { display_usage(); printf("Error: input 'zx' must be greater than or equal to one.\n\n"); return EXIT_FAILURE; }
  if(zy < 1.) { display_usage(); printf("Error: input 'zy' must be greater than or equal to one.\n\n"); return EXIT_FAILURE; }
  if(cx < 0) { display_usage(); printf("Error: input 'cx' must be greater than or equal to zero.\n\n"); return EXIT_FAILURE; }
  if(cy < 0) { display_usage(); printf("Error: input 'cy' must be greater than or equal to zero.\n\n"); return EXIT_FAILURE; }
  if(p <= 0) { display_usage(); printf("Error: input 'p' must be greater than or equal to one.\n\n"); return EXIT_FAILURE; }
  if(q < 1 || q >= p) { display_usage(); printf("Error: input 'q' must be greater than in the range [1,p-1].\n\n"); return EXIT_FAILURE; }
  
  /*****************************************************************/
  /* load input TIFF image & perform dimensions consistency checks */
  /*****************************************************************/
  if(EXIT_FAILURE == load_monopage_tiff_into_fftw_complex(&utmp,&Nx,&Ny,fname_in,0)) {
    printf("Error: failed to open input image '%s'\n",fname_in);
    return EXIT_FAILURE;
  }
  
  if(Nx <= 1) {
    printf("Error: the width of the input image 'in' must be at least 2\n\n");
    fftw_free(utmp); 
    return EXIT_FAILURE; 
  } 
  if(Ny <= 1) {
    printf("Error: the height of the input image 'in' must be at least 2\n\n");
    fftw_free(utmp); 
    return EXIT_FAILURE; 
  } 
  
  /**************************************/
  /* load the sequence of displacements */
  /**************************************/
  getasciitranslations(&dx,&dy,&ntranslations,fname_T,vflag);
  nimages = ntranslations;
  
  /*********************************************************************/
  /* CORE OF THE ROUTINE : compute realistic low-resolution stack (u0) */
  /* from the provided input HR image                                  */
  /*********************************************************************/
  gendataset(utmp,&u0,&ref,dx,dy,&Nx,&Ny,&nx,&ny,nimages,&zx,&zy,cx,cy,p,q,fname_ref!=NULL,vflag);
  
  /**************************************/
  /* Save output (multipage) TIFF image */
  /**************************************/
  if(vflag) {
    printf("Save output images\n");
    printf("==================\n");
  }
  if(ref) {
    if(EXIT_FAILURE == save_monopage_striped_tiff_from_fftw_complex(fname_ref,ref,Nx,Ny,0,type,"w",NULL,NULL)) {
      printf("Error: failed to save output image '%s' (precision %s)\n",fname_ref,((NULL == datatype)?"float":datatype));
      for(k=0;k<nimages;k++) fftw_free(u0[k]);
      fftw_free(ref); fftw_free(utmp);
      free(u0); free(dx); free(dy);
      return EXIT_FAILURE;
    }
    if(vflag) {
      printf("save reference high-resolution image '%s': done\n",fname_ref);
    }
  }
  for(k=0;k<nimages;k++) {
    if(EXIT_FAILURE == save_monopage_striped_tiff_from_fftw_complex(fname_out,u0[k],nx,ny,0,type,((k==0)?"w":"a"),NULL,NULL)) {
      printf("Error: failed to save output image '%s'\n",fname_out);
      for(k=0;k<nimages;k++) fftw_free(u0[k]);
      if(ref) fftw_free(ref);
      fftw_free(utmp);
      free(u0); free(dx); free(dy);
      return EXIT_FAILURE;
    }
    if(vflag) {
      printf("\33[2K\rsave low-resolution stack '%s' page %d/%d (precision %s): done",fname_out,k+1,nimages,((NULL == datatype)?"float":datatype));
      fflush(stdout);
    }
  }
  if(vflag) printf("\n\n");
  
  /***************/
  /* free memory */
  /***************/
  for(k=0;k<nimages;k++) { fftw_free(u0[k]); }
  if(ref) fftw_free(ref);
  free(u0);
  fftw_free(utmp);
  free(dx);
  free(dy);
  
  return EXIT_SUCCESS;
}

File: matlab/src/gendataset.m
function [u0,ref] = gendataset(in,T,zx,zy,varargin)
%%
% Usage: [u0,ref] = gendataset(in,T,zx,zy,Name,Value)
%
% Input(s)/Output(s):
%
%   in : (input matrix of double) input image
%   T  : (input 2-columns matrix of double) translation vector, must have
%        exactly two components, i.e., T = [tx,ty], where tx denotes the
%        horizontal component of the translation vector, and ty the
%        vertical component of the translation vector
%   zx : (scalar >= 1) subsampling factor along the horizontal direction
%   zy : (scalar >= 1) subsampling factor along the vertical direction
%   u0 : (output hypermatrix or matrix of double) output low-resolution
%        image(s) 
%   ref: (output matrix of double) output high-resolution reference image
%        (corresponds to a cropping of in)
%
% Optional Name-Value pair arguments:
%
%   ['complex',c] : (scalar logical, default c = false), set c = true to
%                   return the complex signal (do not take the real part)
%
%   ['cx',cx]     : (scalar with integer value, default cx = 10), maximal
%                   number of columns to be discarded from the input image
%                   during the first cropping step
%
%   ['cy',cy]     : (scalar with integer value, default cy = 10), maximal
%                   number of row to be discarded from the input image
%                   during the first cropping step
%
%   ['p',p]       : (scalar with integer value, default p = 5) he final
%                   high-resolution ROI size will be equal to (q/p) * the
%                   dimensions of the cropped HR image
%
%   ['q',q]       : (scalar with integer value, default q = 4), see above
%
% Description: Compute a realistic low-resolution stack from a high
%              resolution image (avoiding periodic-like boundaries).

%% Control number of inputs
if(nargin < 4)
    help gendataset;
    error('Incorrect number of input(s)');
end

%% parser (consistency checks are done after, to allow precise error messages)
P = inputParser;
P.addRequired('in');
P.addRequired('T');
P.addRequired('zx');
P.addRequired('zy');
P.addParameter('cx',10);
P.addParameter('cy',10);
P.addParameter('p',5);
P.addParameter('q',4);
P.addParameter('complex',false);
parse(P,in,T,zx,zy,varargin{:});
cx = P.Results.cx;
cy = P.Results.cy;
p = P.Results.p;
q = P.Results.q;
cmode = P.Results.complex;

%% consistency checks
% input in (matrix of double real numbers)
if(~ismatrix(in) || isscalar(in) || ~isreal(in))
    help gendataset;
    error('input image in must be a (non-scalar) matrix of double real numbers');
end
% input T (matrix of two double real numbers)
if(~isreal(T) || size(T,2) ~= 2)
    help gendataset;
    error('input translation vector T must have exactly two columns of double real numbers');
end
% input zx (positive scalar >= 1)
if(~isreal(zx) || ~isscalar(zx) || zx < 1)
    help gendataset;
    error('input zx must be a scalar number >= 1');
end
% input zy (positive scalar >= 1)
if(~isreal(zy) || ~isscalar(zy) || zy < 1)
    help gendataset;
    error('input zy must be a scalar number >= 1');
end
% input cx (scalar in [0,size(in,2)] without decimal part)
if(~isreal(cx) || ~isscalar(cx) || cx ~= floor(cx) || cx < 0 || cx > size(in,2))
    help gendataset;
    error('input cx must be a scalar number in [0,size(in,2)] without decimal part');
end
% input cy (scalar in [0,size(in,1)] without decimal part)
if(~isreal(cy) || ~isscalar(cy) || cy ~= floor(cy) || cy < 0 || cy > size(in,1))
    help gendataset;
    error('input cy must be a scalar number in [0,size(in,1)] without decimal part');
end
% input p (scalar >= 2 without decimal part)
if(~isreal(p) || ~isscalar(p) || p ~= floor(p) || p < 2)
    help gendataset;
    error('input p must be a scalar number >= 1 without decimal part');
end
% input q (positive scalar in [1,p-1], without decimal part)
if(~isreal(q) || ~isscalar(q) || q ~= floor(q) || q < 1 || q >= p)
    help gendataset;
    error('input q must be a scalar number in [1, p-1] without decimal part');
end
% input 'complex'
if(~islogical(cmode) || ~isscalar(cmode))
    help gendataset;
    error('you must set c=true or c=false for the optional Name-Value pair argument [''complex'',c]');
end

%% CORE OF THE MODULE: compute a realistic low-resolution stack

% first cropping step along the X & Y axes (we will remove up to cx columns
% and cy rows from the input HR image in order fulfill as best as possible
% the zx and zy requirements)
zx_tmp = zx; zy_tmp = zy; 
[N0, M0] = size(in);
Ax = floor((M0 - (0:cx)) / p);
Ay = floor((N0 - (0:cy)) / p);
ax = round(Ax/zx_tmp);
ay = round(Ay/zy_tmp);
Cx = (zx_tmp - Ax./ax).^2;
Cy = (zy_tmp - Ay./ay).^2;
idx = find(Cx == min(Cx), 1);
idy = find(Cy == min(Cy), 1);
Ax = Ax(idx);
ax = ax(idx);
Ay = Ay(idy);
ay = ay(idy);
zx = Ax/ax;
zy = Ay/ay;
if(zx_tmp ~= zx)
    warning("zx has been changed into %.17g (instead of zx=%g) to allow integer width for the output image",zx,zx_tmp);
end
if(zy_tmp ~= zy) 
    warning("zy has been changed into %.17g (instead of zy=%g) to allow integer height for the output image",zy,zy_tmp);
end
M0prime = Ax * p;
N0prime = Ay * p;
X0 = floor((M0 - M0prime)/2);
Y0 = floor((N0 - N0prime)/2);
ref = in(Y0+(1:N0prime), X0+(1:M0prime));

% perform the periodic + smooth decomposition of the cropped HR image
[P,S] = perdecomp(ref);

% generate (unrealistic) LR stack: use the simulator for the periodic
% component and use bilinear interpolation for the smooth component
m = q*ax;
n = q*ay;
M = q*Ax;
N = q*Ay;
X0 = floor((M0prime - M)/2);
Y0 = floor((N0prime - N)/2);
x0 = X0/zx;
y0 = Y0/zy;
k0 = floor(x0);
l0 = floor(y0);
dx = x0 - k0;
dy = y0 - l0;
T2 = T+ [dx, dy];
m0 = ax*p;
n0 = ay*p;
p0 = simulator(P, T2, m0, n0);
[x, y] = meshgrid(0:m0-1, 0:n0-1);
s0 = zeros(n0,m0,size(T,1));
for id = 1:size(T,1)
    s0(:,:,id) = interp2(S, 1 + zx*(x+T2(id,1)), 1 + zy*(y+T(id,2)), 'linear', 0);
end
u0 = p0 + s0;

% crop the generated stack to get rid of unrealistic boundaries and compute
% the associated HR reference image
ref = ref(Y0+(1:N), X0+(1:M));
u0 = u0(l0+(1:n), k0+(1:m), :);
if(~cmode); u0 = real(u0); end

end